File: backend/services/practice_service.py
import hashlib
import json
from datetime import UTC, datetime, timedelta
from uuid import uuid4

from sqlalchemy import select
from sqlalchemy.orm import Session

from backend.db import models
from backend.schemas import (
    ErrorBatchRequest,
    ErrorBatchResponse,
    PracticeSessionOut,
    PracticeSessionRequest,
    PracticeSessionResponse,
)
from backend.services import profile_service


class PracticeError(Exception):
    pass


def _id(prefix: str) -> str:
    return f"{prefix}_{uuid4().hex[:12]}"


def _hash_question(text: str) -> str:
    return hashlib.sha256(text.strip().encode("utf-8")).hexdigest()


def create_practice_session(db: Session, payload: PracticeSessionRequest) -> PracticeSessionResponse:
    if payload.correct_count > payload.question_count:
        raise PracticeError("correct_count cannot exceed question_count")
    task = db.get(models.DailyTask, payload.daily_task_id)
    module = db.get(models.Module, payload.module_id)
    if task is None or module is None:
        raise PracticeError("Task or module not found")
    if task.module_id != module.id:
        raise PracticeError("Task does not belong to module")

    ended_at = datetime.now(UTC).replace(tzinfo=None)
    started_at = ended_at - timedelta(seconds=payload.duration_seconds)
    session = models.PracticeSession(
        id=_id("session"),
        daily_task_id=task.id,
        module_id=module.id,
        started_at=started_at,
        ended_at=ended_at,
        question_count=payload.question_count,
        correct_count=payload.correct_count,
        accuracy=round(payload.correct_count / payload.question_count, 4),
        duration_seconds=payload.duration_seconds,
        tags_json=json.dumps(payload.tags, ensure_ascii=False),
    )
    db.add(session)
    task.actual_minutes = max(task.actual_minutes, round(payload.duration_seconds / 60))
    if task.type in ("practice", "mock_exam"):
        task.status = "completed"
    profile_service.recalculate_module(db, module)
    goal = module.track.goal
    profile = profile_service.recalculate_profile(db, goal)
    db.commit()
    db.refresh(session)
    db.refresh(task)
    db.refresh(module)
    return PracticeSessionResponse(
        session=PracticeSessionOut(
            id=session.id,
            daily_task_id=session.daily_task_id,
            module_id=session.module_id,
            started_at=session.started_at,
            ended_at=session.ended_at,
            question_count=session.question_count,
            correct_count=session.correct_count,
            accuracy=session.accuracy,
            duration_seconds=session.duration_seconds,
            tags=json.loads(session.tags_json),
        ),
        task=task,
        module=module,
        profile=profile_service.serialize_profile(profile),
    )


def create_error_batch(db: Session, payload: ErrorBatchRequest) -> ErrorBatchResponse:
    session = db.get(models.PracticeSession, payload.practice_session_id)
    if session is None:
        raise PracticeError("Practice session not found")

    created = 0
    skipped = 0
    touched_modules: set[str] = set()
    for item in payload.errors:
        question_hash = _hash_question(item.question_text)
        existing = db.scalar(
            select(models.ErrorRecord).where(
                models.ErrorRecord.practice_session_id == payload.practice_session_id,
                models.ErrorRecord.question_hash == question_hash,
            )
        )
        if existing is not None:
            skipped += 1
            continue
        db.add(
            models.ErrorRecord(
                id=_id("error"),
                practice_session_id=payload.practice_session_id,
                module_id=item.module_id,
                question_hash=question_hash,
                question_text=item.question_text,
                user_answer=item.user_answer,
                correct_answer=item.correct_answer,
                explanation=item.explanation,
                tags_json=json.dumps(item.tags, ensure_ascii=False),
                recorded_at=datetime.now(UTC).replace(tzinfo=None),
                reviewed_count=0,
                mastered=False,
            )
        )
        touched_modules.add(item.module_id)
        created += 1

    goal = None
    for module_id in touched_modules:
        module = db.get(models.Module, module_id)
        if module is not None:
            goal = module.track.goal
    profile = profile_service.recalculate_profile(db, goal) if goal is not None else None
    db.commit()
    return ErrorBatchResponse(created_count=created, skipped_count=skipped, profile=profile_service.serialize_profile(profile))

File: server/services/practice-service.ts
import { createHash, randomUUID } from "crypto";
import { db } from "../db.js";
import type {
  PracticeSessionRequest,
  PracticeSessionResponse,
  ErrorBatchRequest,
  ErrorBatchResponse,
  PracticeSessionRow,
  DailyTaskRow,
  DailyTaskOut,
  ModuleRow,
  ModuleOut,
} from "../types.js";
import {
  recalculateModule,
  recalculateProfile,
  serializeProfile,
} from "./profile-service.js";

export class PracticeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PracticeError";
  }
}

function genId(prefix: string): string {
  return `${prefix}_${randomUUID().replace(/-/g, "").slice(0, 12)}`;
}

function hashQuestion(text: string): string {
  return createHash("sha256").update(text.trim()).digest("hex");
}

export function createPracticeSession(
  data: PracticeSessionRequest,
): PracticeSessionResponse {
  if (data.correct_count > data.question_count) {
    throw new PracticeError("correct_count cannot exceed question_count");
  }

  const task = db
    .prepare("SELECT * FROM daily_tasks WHERE id = ?")
    .get(data.daily_task_id) as DailyTaskRow | undefined;
  const mod = db
    .prepare("SELECT * FROM modules WHERE id = ?")
    .get(data.module_id) as ModuleRow | undefined;

  if (!task || !mod) {
    throw new PracticeError("Task or module not found");
  }
  if (task.module_id !== mod.id) {
    throw new PracticeError("Task does not belong to module");
  }

  const endedAt = new Date();
  const startedAt = new Date(
    endedAt.getTime() - data.duration_seconds * 1000,
  );

  const sessionId = genId("session");
  const accuracy =
    Math.round(
      (data.correct_count / data.question_count) * 10000,
    ) / 10000;

  db.prepare(
    `INSERT INTO practice_sessions
       (id, daily_task_id, module_id, started_at, ended_at,
        question_count, correct_count, accuracy, duration_seconds, tags_json)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
  ).run(
    sessionId,
    task.id,
    mod.id,
    startedAt.toISOString(),
    endedAt.toISOString(),
    data.question_count,
    data.correct_count,
    accuracy,
    data.duration_seconds,
    JSON.stringify(data.tags ?? []),
  );

  // Update task actual_minutes and status in one query
  const newActualMinutes = Math.max(
    task.actual_minutes,
    Math.round(data.duration_seconds / 60),
  );
  let taskStatus = task.status;
  if (task.type === "practice" || task.type === "mock_exam") {
    taskStatus = "completed";
  }

  db.prepare(
    "UPDATE daily_tasks SET actual_minutes = ?, status = ? WHERE id = ?",
  ).run(newActualMinutes, taskStatus, task.id);

  // Recalculate module proficiency, then profile for the owning goal
  recalculateModule(mod.id);

  const track = db
    .prepare("SELECT goal_id FROM tracks WHERE id = ?")
    .get(mod.track_id) as { goal_id: string } | undefined;
  if (!track) {
    throw new PracticeError("Track not found");
  }
  const profileRow = recalculateProfile(track.goal_id);

  // Read back updated rows (matches Python db.refresh)
  const updatedSession = db
    .prepare("SELECT * FROM practice_sessions WHERE id = ?")
    .get(sessionId) as PracticeSessionRow;
  const updatedTask = db
    .prepare("SELECT * FROM daily_tasks WHERE id = ?")
    .get(task.id) as DailyTaskRow;
  const updatedModule = db
    .prepare("SELECT * FROM modules WHERE id = ?")
    .get(mod.id) as ModuleRow;

  return {
    session: {
      id: updatedSession.id,
      daily_task_id: updatedSession.daily_task_id,
      module_id: updatedSession.module_id,
      started_at: updatedSession.started_at,
      ended_at: updatedSession.ended_at,
      question_count: updatedSession.question_count,
      correct_count: updatedSession.correct_count,
      accuracy: updatedSession.accuracy,
      duration_seconds: updatedSession.duration_seconds,
      tags: JSON.parse(updatedSession.tags_json) as string[],
    },
    task: updatedTask as DailyTaskOut,
    module: updatedModule as ModuleOut,
    profile: serializeProfile(profileRow),
  };
}

export function createErrorBatch(
  data: ErrorBatchRequest,
): ErrorBatchResponse {
  const session = db
    .prepare("SELECT * FROM practice_sessions WHERE id = ?")
    .get(data.practice_session_id) as PracticeSessionRow | undefined;
  if (!session) {
    throw new PracticeError("Practice session not found");
  }

  let created = 0;
  let skipped = 0;
  const touchedModules = new Set<string>();

  for (const item of data.errors) {
    const qHash = hashQuestion(item.question_text);
    const existing = db
      .prepare(
        "SELECT id FROM error_records WHERE practice_session_id = ? AND question_hash = ?",
      )
      .get(data.practice_session_id, qHash) as { id: string } | undefined;

    if (existing) {
      skipped++;
      continue;
    }

    db.prepare(
      `INSERT INTO error_records
         (id, practice_session_id, module_id, question_hash, question_text,
          user_answer, correct_answer, explanation, tags_json, recorded_at,
          reviewed_count, mastered)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    ).run(
      genId("error"),
      data.practice_session_id,
      item.module_id,
      qHash,
      item.question_text,
      item.user_answer,
      item.correct_answer,
      item.explanation ?? "",
      JSON.stringify(item.tags ?? []),
      new Date().toISOString(),
      0,
      0,
    );

    touchedModules.add(item.module_id);
    created++;
  }

  // Find the goal through any touched module (matches Python behavior)
  let goalId: string | null = null;
  for (const moduleId of touchedModules) {
    const row = db
      .prepare(
        `SELECT t.goal_id FROM modules m
         JOIN tracks t ON m.track_id = t.id
         WHERE m.id = ?`,
      )
      .get(moduleId) as { goal_id: string } | undefined;
    if (row) {
      goalId = row.goal_id;
    }
  }

  const profile =
    goalId !== null
      ? serializeProfile(recalculateProfile(goalId))
      : null;

  return {
    created_count: created,
    skipped_count: skipped,
    profile,
  };
}
